pkg/storage/unified/search/document.go

```diff
@@ -0,0 +1,167 @@
+	"github.com/grafana/grafana/pkg/services/store/kind/dashboard"
+	"github.com/grafana/grafana/pkg/storage/unified/resource"
+)
+
+type DocumentBuilderProvider interface {
+	// The list returned here defines the set of resource kinds we know about and how to
+	// convert them to documents.  Long term this will likely need to to understand
+	// the "app manifest" that would includes declarative way to identify searchable fields
+	GetDocumentBuilders(ctx context.Context) ([]resource.DocumentBuilderInfo, error)
+}
+
+// Replaced in enterprise with a version that includes stats
+func ProvideBuilders() DocumentBuilderProvider {
+	return &standardDocumentProvider{}
+}
+
+type standardDocumentProvider struct{}
+
+type defaultDocumentBuilder struct{}
+
+var (
+	_ resource.DocumentBuilder = &defaultDocumentBuilder{}
+	_ DocumentBuilderProvider  = &standardDocumentProvider{}
+)
+
+func (p *standardDocumentProvider) GetDocumentBuilders(ctx context.Context) ([]resource.DocumentBuilderInfo, error) {
+	return []resource.DocumentBuilderInfo{
+		{
+			Builder: &defaultDocumentBuilder{},
+		},
+		{
+			GroupResource: schema.GroupResource{
+				Group:    "dashboard.grafana.app",
+				Resource: "dashboards",
+			},
+
+			// This is a dummy example, and will need resolver setup for enterprise stats and and (eventually) data sources
+			Namespaced: func(ctx context.Context, namespace string, blob resource.BlobSupport) (resource.DocumentBuilder, error) {
+				lookup := dashboard.CreateDatasourceLookup([]*dashboard.DatasourceQueryResult{
+					// TODO, query data sources
+				})
+				return &DashboardDocumentBuilder{
+					Namespace:        namespace,
+					DatasourceLookup: lookup,
+					Stats:            nil, // loaded in enterprise
+					Blob:             blob,
+				}, nil
+			},
+		},
+	}, nil
+}
+
+func (*defaultDocumentBuilder) BuildDocument(_ context.Context, key *resource.ResourceKey, rv int64, value []byte) (resource.IndexableDocument, error) {
+	tmp := &unstructured.Unstructured{}
+	err := tmp.UnmarshalJSON(value)
+	if err != nil {
+		return nil, err
+	}
+
+	obj, err := utils.MetaAccessor(tmp)
+	if err != nil {
+		return nil, err
+	}
+
+	doc := &StandardDocumentFields{}
+	doc.Load(key, rv, obj)
+
+	doc.Title = obj.FindTitle(doc.Name)
+	doc.ByteSize = len(value)
+
+	return doc, nil
+}
+
+// This is common across all resources
+type StandardDocumentFields struct {
+	// unique ID across everything (group+resource+namespace+name)
+	ID string `json:"id"`
+	RV int64  `json:"rv"`
+
+	Group     string `json:"group"`
+	Resource  string `json:"resource"`
+	Namespace string `json:"namespace"`
+	Name      string `json:"name"`
+
+	Folder      string   `json:"folder,omitempty"`
+	Title       string   `json:"title,omitempty"`
+	Description string   `json:"description,omitempty"`
+	Tags        []string `json:"tags,omitempty"`
+	ByteSize    int      `json:"byte_size,omitempty"`
+
+	// Standard k8s style labels
+	Labels map[string]string `json:"labels,omitempty"`
```

Comment:
whats the difference between labels and tags?

Reply:
I suppose what I meant is more along the lines of "what kind of data is meant to be stored in these"?

Wasn't sure if they're meant to hold data specific to the resource's domain, or if they're meant to hold something else like k8s metadata or something 👍